File: src-front/adapters/shared/ai-providers.ts
// AI Providers Commands
import type {
  AiProvidersResponse,
  UpdateProviderSettingsRequest,
  SetDefaultProviderRequest,
  ListModelsResponse,
} from "@/lib/types";

import { invoke } from "./platform";

/**
 * Get all AI providers merged with user settings.
 * Returns catalog data merged with user overrides and computed hasApiKey flag.
 */
export const getAiProviders = async (): Promise<AiProvidersResponse> => {
  return invoke<AiProvidersResponse>("get_ai_providers");
};

/**
 * Update settings for a specific AI provider.
 */
export const updateAiProviderSettings = async (
  request: UpdateProviderSettingsRequest,
): Promise<void> => {
  return invoke<void>("update_ai_provider_settings", { request });
};

/**
 * Set or clear the default AI provider.
 */
export const setDefaultAiProvider = async (request: SetDefaultProviderRequest): Promise<void> => {
  return invoke<void>("set_default_ai_provider", { request });
};

/**
 * List available models from a provider.
 * Fetches models from the provider's API using backend-stored secrets.
 */
export const listAiModels = async (providerId: string): Promise<ListModelsResponse> => {
  return invoke<ListModelsResponse>("list_ai_models", { providerId });
};

File: apps/tauri/src/commands/ai_providers.rs
use std::sync::Arc;

use tauri::State;
use wealthfolio_ai::{
    AiProvidersResponse, ListModelsResponse, ProviderApiError, SetDefaultProviderRequest,
    UpdateProviderSettingsRequest,
};

use crate::context::ServiceContext;

use super::error::CommandResult;

#[tauri::command]
pub async fn get_ai_providers(
    context: State<'_, Arc<ServiceContext>>,
) -> CommandResult<AiProvidersResponse> {
    Ok(context.ai_provider_service().get_ai_providers()?)
}

#[tauri::command]
pub async fn update_ai_provider_settings(
    context: State<'_, Arc<ServiceContext>>,
    request: UpdateProviderSettingsRequest,
) -> CommandResult<()> {
    context
        .ai_provider_service()
        .update_provider_settings(request)
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn set_default_ai_provider(
    context: State<'_, Arc<ServiceContext>>,
    request: SetDefaultProviderRequest,
) -> CommandResult<()> {
    context
        .ai_provider_service()
        .set_default_provider(request)
        .await?;
    Ok(())
}

/// List available models from a provider.
/// Fetches models from the provider's API using backend-stored secrets.
/// Frontend never needs to send API keys - they are retrieved internally.
#[tauri::command]
pub async fn list_ai_models(
    context: State<'_, Arc<ServiceContext>>,
    provider_id: String,
) -> Result<ListModelsResponse, ProviderApiError> {
    context
        .ai_provider_service()
        .list_models(&provider_id)
        .await
}
